### PR TITLE
planner: add memory control function for the new plan cache LRU

### PR DESCRIPTION
**Pull request by fzzf678**
### What problem does this PR solve?

Use `quota` and  `guard` control the memory, make it's behavior identical to the simple LRU

Issue Number: ref #36598

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/plan_cache_lru.go [hunk 1]
```diff
@@ -173,3 +181,16 @@ func (l *LRUPlanCache) removeFromBucket(element *list.Element) {
 	bucket := l.buckets[string(hack.String(element.Value.(*planCacheEntry).PlanKey.Hash()))]
 	delete(bucket, element)
 }
+
+// memoryControl control the memory by quota and guard
+func (l *LRUPlanCache) memoryControl() {
+	if l.quota == 0 {
```
Comment (qw4990) [inline planner/core/plan_cache_lru.go:187]: How about `if l.quota==0 || l.guard==0`?

File: planner/core/plan_cache_lru_test.go [hunk 2]
```diff
@@ -69,13 +70,16 @@ func pickFromBucket(bucket map[*list.Element]struct{}, ptypes []*types.FieldType
 }
 
 func TestLRUPCPut(t *testing.T) {
+	maxMem, err := memory.MemTotal()
+	require.NoError(t, err)
+
 	// test initialize
-	lruA, errA := NewLRUPlanCache(0, pickFromBucket)
+	lruA, errA := NewLRUPlanCache(0, 0, maxMem, pickFromBucket)
```
Comment (qw4990) [inline planner/core/plan_cache_lru_test.go:77]: Better to use `0` instead of `maxMem`.

**Review by qw4990 — APPROVED** [review]

File: planner/core/plan_cache_lru.go [hunk 3]
```diff
@@ -43,11 +44,15 @@ type LRUPlanCache struct {
 	pickFromBucket func(map[*list.Element]struct{}, []*types.FieldType) (*list.Element, bool)
 	// onEvict will be called if any eviction happened, only for test use now
 	onEvict func(kvcache.Key, kvcache.Value)
+
+	// 0 indicates no quota
+	quota uint64
+	guard float64
```
Comment (Reminiscent) [inline planner/core/plan_cache_lru.go:50]: Are there any system variables to control this?

Comment (fzzf678) [reply planner/core/plan_cache_lru.go:50]: > Are there any system variables to control this?

yes, to see  `PreparedPlanCacheMaxMemory` and `PreparedPlanCacheMemoryGuardRatio` in `tidb_vars.go`

Comment (fzzf678) [reply planner/core/plan_cache_lru.go:50]: > > Are there any system variables to control this?
> 
> yes, to see `PreparedPlanCacheMaxMemory` and `PreparedPlanCacheMemoryGuardRatio`

sry，`PreparedPlanCacheMaxMemory` is not a system variable,  it is localized, use server-memory-quota as max-plan-cache-memory

**Review by Reminiscent — APPROVED** [review]

**Pull request merged**